File: api/analyze.js
// 文件名: api/analyze.js

const csv = require('csv-parser');
const stream = require('stream');
const _ = require('lodash');

// Vercel Serverless Function (Node.js/Express 格式)
module.exports = async (req, res) => {
    // 設置 CORS 標頭
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

    if (req.method === 'OPTIONS') {
        return res.status(200).end();
    }

    if (req.method !== 'POST') {
        return res.status(405).json({ success: false, message: 'Method Not Allowed' });
    }

    try {
        const { data: rawData, clean_missing, normalize_data, sort_sales } = req.body;
        if (!rawData) {
            return res.status(400).json({ success: false, message: '數據為空。' });
        }

        const data = [];
        const bufferStream = new stream.PassThrough();
        bufferStream.end(rawData);

        // 1. 數據解析 (CSV 解析)
        await new Promise((resolve, reject) => {
            bufferStream
                .pipe(csv())
                .on('data', (row) => {
                    // 將 Sales 欄位轉換為數字，如果轉換失敗則設為 NaN
                    row.Sales = parseFloat(row.Sales);
                    data.push(row);
                })
                .on('end', () => resolve())
                .on('error', (err) => reject(err));
        });

        let processedData = data;

        // 2. 數據處理核心邏輯

        // 2.1 清理缺失值
        if (clean_missing) {
            // 移除 Sales 為 NaN 的行
            processedData = processedData.filter(row => !isNaN(row.Sales));
        }

        const salesValues = processedData.map(row => row.Sales);

        // 2.2 數據標準化 (Z-Score)
        if (normalize_data && salesValues.length > 0) {
            const mean = _.mean(salesValues);
            // 計算標準差 (使用樣本標準差 N)
            const std = Math.sqrt(_.sum(salesValues.map(v => Math.pow(v - mean, 2))) / salesValues.length); 

            if (std > 0) {
                processedData = processedData.map(row => {
                    row.Sales = _.round((row.Sales - mean) / std, 4);
                    return row;
                });
            } else {
                // 標準差為 0，所有值都一樣，Z-Score 為 0
                processedData = processedData.map(row => { row.Sales = 0; return row; });
            }
        }

        // 2.3 排序
        if (sort_sales) {
            // lodash 的 sortBy 預設是升序，使用 reverse 變成降序
            processedData = _.sortBy(processedData, ['Sales']).reverse(); 
        }

        // 3. 生成統計摘要 (模擬 Pandas describe())
        const stats = {};
        const validSales = processedData.map(row => row.Sales).filter(v => !isNaN(v));

        if (validSales.length > 0) {
            const mean = _.mean(validSales);
            const std = Math.sqrt(_.sum(validSales.map(v => Math.pow(v - mean, 2))) / validSales.length);
            
            stats.Sales = {
                count: validSales.length,
                mean: mean,
                std: std,
                min: _.min(validSales),
                max: _.max(validSales)
            };
        }

        // 4. 返回結果
        return res.status(200).json({
            success: true,
            processed_data: processedData,
            statistics: stats,
            message: "Node.js 數據分析成功完成。"
        });

    } catch (error) {
        console.error('Node.js 處理數據時發生錯誤:', error);
        return res.status(500).json({ success: false, message: `後端處理數據時發生錯誤: ${error.message}` });
    }
};

File: api/analyze.py
# 文件名: api/analyze.py

import json
import pandas as pd
import io
from scipy.stats import zscore # 引入 scipy 進行標準化 (通常 Pandas 自帶或依賴 numpy/scipy)

def handler(event, context):
    """
    Vercel Serverless Function 的入口點。
    處理前端傳來的數據和處理選項。
    """
    
    # 設置 CORS 標頭，允許前端調用
    headers = {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Methods": "POST, OPTIONS",
        "Access-Control-Allow-Headers": "Content-Type"
    }
    
    # 處理 OPTIONS 請求 (CORS 預檢)
    if event['httpMethod'] == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': headers,
            'body': ''
        }
        
    if event['httpMethod'] != 'POST':
        return {
            'statusCode': 405,
            'headers': headers,
            'body': json.dumps({"success": False, "message": "Method Not Allowed"})
        }

    try:
        # 1. 解析前端傳來的 JSON 數據
        body = json.loads(event['body'])
        raw_data = body.get('data', '')
        clean_missing = body.get('clean_missing', False)
        normalize_data = body.get('normalize_data', False)
        sort_sales = body.get('sort_sales', False)
        
        if not raw_data:
            raise ValueError("數據為空。")

        # 2. 將 CSV 字符串讀取到 Pandas DataFrame
        df = pd.read_csv(io.StringIO(raw_data))

        # 3. 數據處理核心邏輯 (依據前端開關)

        # 3.1 清理缺失值
        if clean_missing:
            df.dropna(inplace=True)

        # 3.2 數據標準化 (Z-Score)
        if normalize_data and 'Sales' in df.columns:
            # 確保 'Sales' 是數字類型
            df['Sales'] = pd.to_numeric(df['Sales'], errors='coerce')
            df['Sales_ZScore'] = zscore(df['Sales'])
            # 為了簡潔顯示，這裡我們將 'Sales' 替換為 ZScore
            df['Sales'] = df['Sales_ZScore']
            df.drop(columns=['Sales_ZScore'], inplace=True)


        # 3.3 排序
        if sort_sales and 'Sales' in df.columns:
            df.sort_values(by='Sales', ascending=False, inplace=True)

        # 4. 生成統計摘要
        # 僅對數字欄位進行描述性統計
        stats = df.describe(include=[pd.np.number]).T.to_dict()
        
        # 5. 準備結果
        # 將處理後的 DataFrame 轉換為 JSON 列表，以便前端 JS 處理
        processed_data_list = df.fillna('').to_dict('records')

        response_data = {
            "success": True,
            "processed_data": processed_data_list,
            "statistics": stats,
            "message": "數據分析成功完成。"
        }
        
        return {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps(response_data)
        }

    except Exception as e:
        print(f"處理錯誤: {e}")
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({"success": False, "message": f"伺服器錯誤: {str(e)}"})
        }